core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -0,0 +1,568 @@
+import static java.util.function.Predicate.not;
+
+/**
+ * Rule that transforms selected aggregations from:
+ * <pre> {@code
+ * - Aggregation[key]: sum(aggr1), sum(aggr2), ..., sum(aggrN)
+ *   - Project:
+ *       aggr1 = CASE WHEN col=1 THEN expr ELSE 0
+ *       aggr2 = CASE WHEN col=2 THEN expr ELSE null
+ *       ...
+ *       aggrN = CASE WHEN col=N THEN expr
+ *     - source
+ * }
+ * </pre>
+ * into:
+ * <pre> {@code
+ * - Aggregation[key]: sum(aggr1), sum(aggr2), ..., sum(aggrN)
+ *   - Project:
+ *       aggr1 = CASE WHEN col=1 THEN pre_aggregate
+ *       aggr2 = CASE WHEN col=2 THEN pre_aggregate
+ *       ..
+ *       aggrN = CASE WHEN col=N THEN pre_aggregate
+ *     - Aggregation[key, col]: pre_aggregate = sum(expr)
+ *       - source
+ * }
+ * </pre>
+ */
+public class PreAggregateCaseAggregations
+        implements Rule<AggregationNode>
+{
+    private static final int MIN_AGGREGATION_COUNT = 4;
+    private static final Set<String> ALLOWED_FUNCTIONS = ImmutableSet.of("max", "min", "sum");
+    private static final Capture<ProjectNode> PROJECT_CAPTURE = newCapture();
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(aggregation -> aggregation.getStep() == SINGLE)
+            .with(source().matching(project().capturedAs(PROJECT_CAPTURE)
+                    // prevent rule from looping by ensuring that projection source is not aggregation
+                    .with(source().matching(not(AggregationNode.class::isInstance)))));
+
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public PreAggregateCaseAggregations(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer)
+    {
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isPreAggregateCaseAggregationsEnabled(session);
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        ProjectNode projectNode = captures.get(PROJECT_CAPTURE);
+        Optional<List<CaseAggregation>> aggregationsOptional = extractCaseAggregations(aggregationNode, projectNode, context);
+        if (aggregationsOptional.isEmpty()) {
+            return Result.empty();
+        }
+
+        List<CaseAggregation> aggregations = aggregationsOptional.get();
+        if (aggregations.size() < MIN_AGGREGATION_COUNT) {
+            return Result.empty();
+        }
+
+        Set<Symbol> extraGroupingKeys = aggregations.stream()
+                .flatMap(expression -> SymbolsExtractor.extractUnique(expression.getOperand()).stream())
+                .collect(toImmutableSet());
```

Comment:
I can't find where the code analyzes the shape of the operand expressions. If we want to group the input based on them, we should make sure that they are equality conditions.

Reply:
Operands doesn't really matter as operand column gets preserved. Can be any condition really.
I've also tested on queries like `q50` with conditions like:
```
, "sum"((CASE WHEN (("sr_returned_date_sk" - "ss_sold_date_sk") <= 30) THEN 1 ELSE 0 END)) "30 days"
, "sum"((CASE WHEN (("sr_returned_date_sk" - "ss_sold_date_sk") > 30)
   AND (("sr_returned_date_sk" - "ss_sold_date_sk") <= 60) THEN 1 ELSE 0 END)) "31-60 days"
, "sum"((CASE WHEN (("sr_returned_date_sk" - "ss_sold_date_sk") > 60)
   AND (("sr_returned_date_sk" - "ss_sold_date_sk") <= 90) THEN 1 ELSE 0 END)) "61-90 days"
, "sum"((CASE WHEN (("sr_returned_date_sk" - "ss_sold_date_sk") > 90)
   AND (("sr_returned_date_sk" - "ss_sold_date_sk") <= 120) THEN 1 ELSE 0 END)) "91-120 days"
, "sum"((CASE WHEN (("sr_returned_date_sk" - "ss_sold_date_sk") > 120) THEN 1 ELSE 0 END)) ">120 days"
```
but I havent seen much improvement when two extra grouping columns are added `sr_returned_date_sk, ss_sold_date_sk`.